packages/backend/src/telemetry/service.ts

```diff
@@ -0,0 +1,147 @@
+
+export interface TelemetryService {
+  getOrCreateMetric(name: string, options?: MetricOptions): Counter
+  getServiceName(): string | undefined
+  getBaseAssetCode(): string
+  getBaseScale(): number
+  convertAmount(
+    convertOptions: Omit<ConvertOptions, 'exchangeRate' | 'destinationAsset'>
+  ): Promise<bigint | ConvertError>
+}
+
+interface TelemetryServiceDependencies extends BaseService {
+  serviceName: string
+  collectorUrls: string[]
+  exportIntervalMillis?: number
+  aseRatesService: RatesService
+  fallbackRatesService: RatesService
+  baseAssetCode: string
+  baseScale: number
+}
+
+export function createTelemetryService(
+  deps: TelemetryServiceDependencies
+): TelemetryService {
+  return new TelemetryServiceImpl(deps)
+}
+
+class TelemetryServiceImpl implements TelemetryService {
+  private serviceName: string
+  private meterProvider?: MeterProvider
+  private fallbackRatesService: RatesService
+  private aseRatesService: RatesService
+
+  private counters = new Map()
+  constructor(private deps: TelemetryServiceDependencies) {
+    diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG)
+    this.serviceName = deps.serviceName
+    this.fallbackRatesService = deps.fallbackRatesService
+    this.aseRatesService = deps.aseRatesService
+
+    if (
+      deps.collectorUrls &&
+      Array.isArray(deps.collectorUrls) &&
+      deps.collectorUrls.length === 0
+    ) {
+      deps.logger.info(
+        'No collector URLs specified, metrics will not be exported'
+      )
+      return
+    }
+
+    this.meterProvider = new MeterProvider({
+      resource: new Resource({ 'service.name': 'RAFIKI_NETWORK' })
+    })
+
+    deps.collectorUrls.forEach((url) => {
+      const metricExporter = new OTLPMetricExporter({
+        url: url
+      })
+
+      const metricReader = new PeriodicExportingMetricReader({
+        exporter: metricExporter,
+        exportIntervalMillis: deps.exportIntervalMillis ?? 15000
+      })
+
+      this.meterProvider?.addMetricReader(metricReader)
+    })
+
+    metrics.setGlobalMeterProvider(this.meterProvider)
+  }
+
+  private createCounter(
+    name: string,
+    options: MetricOptions | undefined
+  ): Counter {
+    const counter = metrics.getMeter('Rafiki').createCounter(name, options)
+    this.counters.set(name, counter)
+    return counter
+  }
+
+  public getOrCreateMetric(name: string, options?: MetricOptions): Counter {
+    const existing = this.counters.get(name)
+    if (existing) {
+      return existing
+    }
+    return this.createCounter(name, options)
+  }
+
+  public async convertAmount(
+    convertOptions: Omit<ConvertOptions, 'exchangeRate'>
+  ) {
+    const destinationAsset = {
+      code: this.deps.baseAssetCode,
+      scale: this.deps.baseScale
+    }
+
+    let converted = await this.aseRatesService.convert({
+      ...convertOptions,
+      destinationAsset
+    })
+    if (typeof converted !== 'bigint' && converted in ConvertError) {
```

Comment:
Kinda just a style thing but if you wanted you could do this:
```ts
    if (isConvertError(converted)) {
      converted = await this.fallbackRatesService.convert(convertOptions)
      if (isConvertError(converted)) {
        this.deps.logger.error(`Unable to convert amount: ${converted}`)
      }
    }
```

After adding this to `rafiki/packages/backend/src/rates/service.ts`
```ts
// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
export const isConvertError = (o: any): o is ConvertError =>
  Object.values(ConvertError).includes(o)
```

This is a pattern we follow a lot of other places but looks like we just didnt add such a util for `ConvertError`. I like it but I guess its kinda personal preference since both narrow the types. Seems like the `&& converted in ConvertError` is not strictly necessary in any case.

Reply:
sure thing, i'm pro typescript guards. Will add, thanks